### PR TITLE
add tests for calendar utils

### PR DESCRIPTION
**Pull request by akpax**
Tests Cover following cases:
- weekends
- weekdays
- holidays

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
+name: test
+
```
Comment (fallonjoey) [inline .github/workflows/build.yml:2]: I will want a demo on gitlab bc Im interested

Global comments:
- This is a good opportunity to try out some other git cli commands. Typically PRs are viewed as final state of code. We are not tracking our work with agile so its fair that this is a big PR. Just a heads up a lot of commits on a PR looks like you did not finish the code before you submitted it (only in a company environment)
- Try getting this down to one single commit by editing the history with git cli commands

File: dags/util/core/alpaca/alpaca_actions.py [hunk 2]
```diff
@@ -30,4 +31,7 @@ def pull_bars(start_date, end_date, header):
     )
     # get bars and convert to pandas_df
     bars_df = stock_client.get_stock_bars(request_params).df
-    return bars_df.reset_index()
+    if len(bars_df) > 0:
+        return bars_df.reset_index()
+    else:
+        return None
```
Comment (fallonjoey) [inline dags/util/core/alpaca/alpaca_actions.py:37]: could save a line of code here and just remove the else statement and return None.

File: pyproject.toml [hunk 3]
```diff
@@ -13,6 +13,11 @@ pandas = "^2.2.2"
 google-cloud = "^0.34.0"
 google-cloud-bigquery = "3.23.0"
 pyarrow = "16.1.0"
+google-cloud-storage = "^2.16.0"
+gcsfs = "^2024.6.0"
+pandas-market-calendars = "^4.4.0"
```
Comment (fallonjoey) [inline pyproject.toml:18]: looks like a cool package. For future reference this is something we could do with datetime native python packages. More packages means more to manage and more that could break

File: tests/test_calendar_utils.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import pytest
+
+from ..dags.util.core.calendar.calendar_util import check_if_exchange_open
+from datetime import date
+
+
+EXCHANGE_CODE = "NYSE"
+
+
+class Test_check_if_exchange_open:
```
Comment (fallonjoey) [inline tests/test_calendar_utils.py:10]: What is the use for putting everything in a class?

File: pytest.ini [hunk 5]
```diff
@@ -0,0 +1,2 @@
+[pytest]
+keepduplicates=false
```
Comment (fallonjoey) [inline pytest.ini:2]: Do we have tests with the same name? If we do, I feel like just changing the name of the test is simpler and cleaner than this config file right? LMK if I am misunderstanding the purpose of this config

**Review by fallonjoey — CHANGES_REQUESTED** [review]
Action items:
1. Consolidate git commit history to a single commit (this is great practice since its a straight forward example and super useful in industry)
2. Demo gitlabs
3. Discuss other comments if your interested but these are not pressing

File: pytest.ini (hunk 5 above)
Comment (akpax) [reply pytest.ini]: I deleted this file. I was having an issue with duplicate tests and this was residual.

File: pyproject.toml (hunk 3 above)
Comment (akpax) [reply pyproject.toml:18]: Can we check market holidays with native packages?

File: dags/util/core/alpaca/alpaca_actions.py (hunk 2 above)
Comment (akpax) [reply dags/util/core/alpaca/alpaca_actions.py:37]: incorporated this change

File: .github/workflows/build.yml (hunk 1 above)
Comment (akpax) [reply .github/workflows/build.yml:2]: I got it down to one commit! Thanks for the tip.

Comment (akpax) [reply .github/workflows/build.yml:2]: Also why do we need a GitLab demo if we are hosting on Git?  I noticed GitLab has some CI/CD workflows that can be done via GUI but im not sure why we would want to switch versioning controls now. Can we discuss?

**Pull request merged**